gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+use crate::StorableMsg;
+use anyhow::anyhow;
+use cid::multihash::Code;
+use cid::multihash::MultihashDigest;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::to_vec;
+use fvm_ipld_encoding::tuple::{Deserialize_tuple, Serialize_tuple};
+use fvm_ipld_hamt::BytesKey;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use ipc_sdk::ValidatorSet;
+use primitives::{TCid, THamt};
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+pub type HashOutput = Vec<u8>;
+const RATIO_NUMERATOR: u16 = 2;
+const RATIO_DENOMINATOR: u16 = 3;
+
+/// Checkpoints propagated from parent to child to signal the "final view" of the parent chain
+/// from the different validators in the subnet.
+#[derive(Clone, Debug, Serialize_tuple, Deserialize_tuple, PartialEq, Eq)]
+pub struct CronCheckpoint {
+    pub epoch: ChainEpoch,
+    pub validators: ValidatorSet,
+    pub top_down_msgs: Vec<StorableMsg>,
+}
+
+impl CronCheckpoint {
+    /// Hash the checkpoint.
+    ///
+    /// To compare the cron checkpoint and ensure they are the same, we need to make sure the
+    /// validators and top_down_msgs are the same. However, the top_down_msgs and validators are vec,
+    /// they may contain the same content, but their orders are different. In this case, we need to
+    /// ensure the same order is maintained in the cron checkpoint submission.
+    ///
+    /// To ensure we have the same consistent output for different submissions, we require:
+    ///     - validators are sorted by `net_addr` in string ascending order
+    ///     - top down messages are sorted by (from, to, nonce) in descending order
+    ///
+    /// Actor will not perform sorting to save gas. Client should do it, actor just check.
+    fn hash(&self) -> anyhow::Result<HashOutput> {
+        // check validators
```

Comment:
This can be removed. As described above we won't report validators in the crone anymore. That being said, this is a really nice approach :raised_hands:

Reply:
But I think we still need to check the top down messages are the same.

Reply:
We don't need validator order.